### PR TITLE
Add ThreadCount type

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

**Comment by wendigo** [comment]
cc @martint @electrum @hashhar

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import static io.airlift.units.Preconditions.checkArgument;
+import static java.lang.Math.min;
+import static java.lang.Math.round;
+import static java.lang.String.format;
+
+public class ThreadsCount
+        implements Comparable<ThreadsCount>
+{
+    private static final String PER_CORE_SUFFIX = "C";
+    private static final Supplier<Integer> AVAILABLE_PROCESSORS = MachineInfo::getAvailablePhysicalProcessorCount;
```
Comment (sopel39) [inline src/main/java/io/airlift/units/ThreadsCount.java:29]: Did you check https://github.com/trinodb/trino/pull/16303#discussion_r1122780546?

Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadsCount.java:29]: Yes.

I've created a Dockerfile that runs a main method added to ThreadsCount:

```
 public static void main(String[] args)
 {
        System.out.println(ThreadsCount.valueOf("1.0C"));
 }
```

Dockerfile:
```
ARG JAVA_VERSION=17
FROM arm64v8/openjdk:${JAVA_VERSION}-ea-jdk

COPY target/units-1.9-SNAPSHOT.jar units.jar
COPY oshi-core.jar oshi-core.jar

CMD java -cp units.jar:oshi-core.jar io.airlift.units.ThreadsCount
```

Then:

```
docker build . -t threads-count
```

And results (8 cores are available to docker, this is ARM so no HT):

Screenshot 2023-03-29 at 14 36 33

```
❯ docker run threads-count
8 threads
❯ docker run --cpus 2 threads-count
2 threads
❯ docker run --cpus 4 threads-count
4 threads
❯ docker run --cpus 2.5 threads-count
3 threads
```

Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadsCount.java:29]: Does it answer your concerns @sopel39 ?

Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadsCount.java:29]: I've copied that code from Trino, so physical cores are taken into the account for Intel

Comment (sopel39) [reply src/main/java/io/airlift/units/ThreadsCount.java:29]: cool!. thanks

**Comment by wendigo** [comment]
@electrum can I make it optional with the default being a /proc/cpuinfo based?

File: src/main/java/io/airlift/units/PowerOfTwo.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Target({METHOD, ANNOTATION_TYPE})
+@Retention(RUNTIME)
+@Documented
+@Constraint(validatedBy = PowerOfTwoValidator.class)
+public @interface PowerOfTwo
```
Comment (dain) [inline src/main/java/io/airlift/units/PowerOfTwo.java:31]: Is this used for `ThreadCount`?  I feel like this and ThreadCount should be separate commits.

**Comment by electrum** [comment]
Optional dependencies are a pain. How do you test that the fallback works? You need a different test suite that doesn’t have the optional dependency on the classpath.

The question is if we can get this from Java without needing a native dependency. If no, then we can evaluate options. If yes, then why make it optional?

**Comment by wendigo** [comment]
@electrum Sure. I'll implement it without deps. Thanks for guidance!

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    @Override
+    public String toString()
+    {
+        if (threadsCount == 1) {
+            return "1 thread";
+        }
+        return format("%d threads", threadsCount);
+    }
+
+    public static final class MachineInfo
```
Comment (losipiuk) [inline src/main/java/io/airlift/units/ThreadsCount.java:148]: `public`?

Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadsCount.java:148]: This class will be dropped entirely

**Comment by wendigo** [comment]
@electrum ptal

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            return Integer.parseInt(value);
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException(format("Cannot parse value '%s' as integer", value), e);
+        }
+    }
+
+    static int nextPowerOfTwo(int x)
+    {
+        if (x == 0) {
```
Comment (electrum) [inline src/main/java/io/airlift/units/ThreadsCount.java:107]: We can base this on Guava's `ceilingPowerOfTwo`, which handles zero automatically:
```java
if (x > (1 << 30)) {
    throw new IllegalStateException("Thread count is too large for next power of two: " + x);
}
return 1 << -Integer.numberOfLeadingZeros(x - 1);

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                            .stream()
+                            .filter(line -> line.matches("^processor\\s+: \\d"))
+                            .count());
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            else {
+                // ARM platforms do not support hyper threading, therefore each logical processor is separate core
```
Comment (electrum) [inline src/main/java/io/airlift/units/ThreadsCount.java:186]: This code might run on any Java platform, not just ARM, so let's remove this comment

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            if (physicalProcessorCount != -1) {
+                return physicalProcessorCount;
+            }
+
+            String osArch = System.getProperty("os.arch");
+            // logical core count (including container cpu quota if there is any)
+            int availableProcessorCount = Runtime.getRuntime().availableProcessors();
+            int totalPhysicalProcessorCount;
+            if ("amd64".equals(osArch) || "x86_64".equals(osArch)) {
+                checkState(exists(CPU_INFO_PATH), "%s does not exist", CPU_INFO_PATH.toString());
```
Comment (electrum) [inline src/main/java/io/airlift/units/ThreadsCount.java:174]: This will fail on non-Linux operating systems. Also, we should ignore any errors reading or parsing this file and fallback to `availableProcessorCount`.

File: src/main/java/io/airlift/units/ThreadsCount.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                try {
+                    totalPhysicalProcessorCount = toIntExact(readAllLines(CPU_INFO_PATH)
```
Comment (electrum) [inline src/main/java/io/airlift/units/ThreadsCount.java:176]: This can use `Files.lines()` which returns a stream

File: src/test/java/io/airlift/units/TestPowerOfTwoValidator.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * limitations under the License.
+ */
+package io.airlift.units;
+
+import org.testng.annotations.Test;
+
+import static io.airlift.testing.ValidationAssertions.assertFailsValidation;
+import static io.airlift.testing.ValidationAssertions.assertValidates;
+
+public class TestPowerOfTwoValidator
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestPowerOfTwoValidator.java:21]: This test class is in the wrong commit. Should be in the one that adds the validator.

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    private void assertBoundedThreadsCount(String value, String min, String max, int expected)
+    {
+        ThreadsCount threadsCount = ThreadsCount.boundedValueOf(value, min, max);
+        assertThat(threadsCount).isEqualTo(new ThreadsCount(expected));
+    }
+
+    private void assertInvalidValue(String value, String expectedMessage)
+    {
+        try {
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:89]:
```java
assertThatThrownBy(() -> ThreadsCount.valueOf(value))
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessage(expectedMessage);

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    @Test
+    public void testParsingMultiplierPerCore()
+    {
+        assertThreadsCount("1C", AVAILABLE_PROCESSORS);
+        assertThreadsCount("0.5 C", AVAILABLE_PROCESSORS / 2);
+        assertThreadsCount("0.2 C", round(AVAILABLE_PROCESSORS / 5.0f));
+        assertThreadsCount("1.5C", round(AVAILABLE_PROCESSORS * 1.5f));
+        assertThreadsCount("2 C", AVAILABLE_PROCESSORS * 2);
+        assertThreadsCount("-0.0001 C", 0);
+        assertInvalidValue("-1C", "Threads count cannot be smaller than 0");
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:46]: Add a test for a value that doesn't parse as a float

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import static java.lang.Math.round;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.testng.Assert.fail;
+
+public class TestThreadsCount
+{
+    public static final int AVAILABLE_PROCESSORS = getAvailablePhysicalProcessorCount();
+
+    @Test
+    public void testParsingIntegerValues()
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:29]: Add tests for
* negative integer
* zero
* value larger than an integer

File: src/test/java/io/airlift/units/TestThreadsCount.java (hunk 10 above)
Comment (electrum) [reply src/test/java/io/airlift/units/TestThreadsCount.java:46]: Add tests for
* "2147483647C" <-- larger than `MAX_INT` if `AVAILABLE_PROCESSORS > 1`
* "3147483648C" <-- larger than `MAX_INT` before multiplying

**Review by electrum — COMMENTED** [review]
Looking good

**Comment by wendigo** [comment]
@electrum all of your comments are addressed now

**Comment by wendigo** [comment]
I've fixed the compilation error on JDK 11 (the last working airlift version is 219, 220 was switched to JDK 17)

**Comment by wendigo** [comment]
@sopel39 ptal

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Test
+    public void testParsingMultiplierPerCorePowerOfTwo()
+    {
+        assertThreadsCountNextPowerOf2("1C", nextPowerOfTwo(AVAILABLE_PROCESSORS));
+        assertThreadsCountNextPowerOf2("0.5C", nextPowerOfTwo(AVAILABLE_PROCESSORS / 2));
+        assertThreadsCountNextPowerOf2("0.2C", nextPowerOfTwo(round(AVAILABLE_PROCESSORS / 5.0f)));
+        assertThreadsCountNextPowerOf2("1.5C", nextPowerOfTwo(round(AVAILABLE_PROCESSORS * 1.5f)));
+        assertThreadsCountNextPowerOf2("2C", nextPowerOfTwo(AVAILABLE_PROCESSORS * 2));
```
Comment (losipiuk) [inline src/test/java/io/airlift/units/TestThreadsCount.java:63]: Nit: as `nextPowerOfTwo` is actally a production code and you are using it here to compute expected value it would be nice to test the `nextPowerOfTwo` itselft too.

**Review by losipiuk — COMMENTED** [review]
LGTM - but I @sopel39 PTAL too.

File: src/test/java/io/airlift/units/TestThreadsCount.java (hunk 12 above)
Comment (wendigo) [reply src/test/java/io/airlift/units/TestThreadsCount.java:63]: Makes sense. I'll add a test for it :)

Comment (wendigo) [reply src/test/java/io/airlift/units/TestThreadsCount.java:63]: Done

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        assertInvalidValue("-1C", "Thread count cannot be negative");
+        assertInvalidValue("-1SC", "Cannot parse value '-1S' as float");
+        assertInvalidValue("2147483647C", "Thread count is greater than 2^32 - 1");
+        assertInvalidValue("3147483648C", "Thread count is greater than 2^32 - 1");
+    }
+
+    @Test
+    public void testParsingMultiplierPerCorePowerOfTwo()
+    {
+        assertThreadsCountNextPowerOf2("1C", nextPowerOfTwo(AVAILABLE_PROCESSORS));
```
Comment (sopel39) [inline src/test/java/io/airlift/units/TestThreadsCount.java:59]: nit: testing machines already probably have pow of two cores, so this test might be noop

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        assertThreadsCountNextPowerOf2("0.5C", nextPowerOfTwo(AVAILABLE_PROCESSORS / 2));
+        assertThreadsCountNextPowerOf2("0.2C", nextPowerOfTwo(round(AVAILABLE_PROCESSORS / 5.0f)));
+        assertThreadsCountNextPowerOf2("1.5C", nextPowerOfTwo(round(AVAILABLE_PROCESSORS * 1.5f)));
+        assertThreadsCountNextPowerOf2("2C", nextPowerOfTwo(AVAILABLE_PROCESSORS * 2));
+        assertThreadsCountNextPowerOf2("0C", 1);
+        assertThreadsCountNextPowerOf2("0.0001C", 1);
```
Comment (sopel39) [inline src/test/java/io/airlift/units/TestThreadsCount.java:65]: Do you think it's find to implicitly round up value provided by user to next power of two? Or maybe we should expect user defined value to be already power of two?

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+import org.testng.annotations.Test;
+
+import static io.airlift.units.ThreadCount.MachineInfo.getAvailablePhysicalProcessorCount;
+import static io.airlift.units.ThreadCount.nextPowerOfTwo;
+import static java.lang.Math.round;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+public class TestThreadsCount
+{
```
Comment (sopel39) [inline src/test/java/io/airlift/units/TestThreadsCount.java:25]: In Trino we have default `taskConcurrency = min(max(nextPowerOfTwo(getAvailablePhysicalProcessorCount()), 2), 32);` (`min/max` boundaries). Will you config work with such boundaries?

File: src/test/java/io/airlift/units/TestThreadsCount.java (hunk 13 above)
Comment (wendigo) [reply src/test/java/io/airlift/units/TestThreadsCount.java:59]: that's why I'm testing nextPowerOfTwo method as a standalone.

My machine has 10 cores :)

File: src/test/java/io/airlift/units/TestThreadsCount.java (hunk 14 above)
Comment (wendigo) [reply src/test/java/io/airlift/units/TestThreadsCount.java:65]: I don't think so. It makes sense when we actually need power of two (some distribution of tasks to the threads using bitmasks or whatever)

File: src/main/java/io/airlift/units/ThreadCount.java [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+        checkArgument(threadCount >= 0, "Thread count cannot be negative");
+        this.threadCount = threadCount;
+    }
+
+    public int getThreadCount()
+    {
+        return threadCount;
+    }
+
+    public ThreadCount nextPowerOfTwo()
```
[the remark this answers is not in the document]
Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadCount.java:49]: It will be harder to use this in Trino (see example that @sopel39 shared): ```taskConcurrency = min(max(nextPowerOfTwo(getAvailablePhysicalProcessorCount()), 2), 32);```

[the remark this answers is not in the document]
Comment (dain) [reply src/main/java/io/airlift/units/ThreadCount.java:49]: > It will be harder to use this in Trino (see example that @sopel39 shared): `taskConcurrency = min(max(nextPowerOfTwo(getAvailablePhysicalProcessorCount()), 2), 32);`

This seems to be more of a complain about the min/max than the power of two call.  I'd switch to Guava's `Ints.constrainToRange` so this code would become:
```java
taskConcurrency = constrainToRange(nextPowerOfTwo(getAvailablePhysicalProcessorCount()), 2, 32)
```
Which I think is totally readable.  BTW, Java 21 has [Math.clamp](https://bugs.openjdk.org/browse/JDK-8301226).

[the remark this answers is not in the document]
Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadCount.java:49]: I'm fine with either keeping this method or dropping it.

[the remark this answers is not in the document]
Comment (wendigo) [reply src/main/java/io/airlift/units/ThreadCount.java]: I've dropped nextPowerOfTwo

**Comment by wendigo** [comment]
@dain @electrum please review. I've dropped `powerOfTwo` on `ThreadCount`

File: src/main/java/io/airlift/units/ThreadCount.java [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        public static int getAvailablePhysicalProcessorCount()
+        {
+            if (physicalProcessorCount != -1) {
+                return physicalProcessorCount;
+            }
+
+            String osArch = System.getProperty("os.arch");
+            // logical core count (including container cpu quota if there is any)
+            int availableProcessorCount = Runtime.getRuntime().availableProcessors();
+            int totalPhysicalProcessorCount;
```
Comment (electrum) [inline src/main/java/io/airlift/units/ThreadCount.java:154]: We could initialize as `availableProcessorCount` here and then the else blocks are not needed

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Test
+    public void testParsingMultiplierPerCore()
+    {
+        assertThreadsCount("1C", AVAILABLE_PROCESSORS);
+        assertThreadsCount("0.5 C", AVAILABLE_PROCESSORS / 2);
+        assertThreadsCount("0.2 C", round(AVAILABLE_PROCESSORS / 5.0f));
+        assertThreadsCount("1.5C", round(AVAILABLE_PROCESSORS * 1.5f));
+        assertThreadsCount("2 C", AVAILABLE_PROCESSORS * 2);
+        assertThreadsCount("-0.0001 C", 0);
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:48]: Relying on this "round small negative to zero" behavior seems strange. Let's add an explicit check that the multiplier is not zero.

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        assertInvalidValue("-1C", "Thread count cannot be negative");
+        assertInvalidValue("-1SC", "Cannot parse value '-1S' as float");
+        assertInvalidValue("2147483647C", "Thread count is greater than 2^32 - 1");
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:51]: Add a check that `AVAILABLE_PROCESSORS > 1` before running this assertion, otherwise it will fail if there is only a single processor.

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    public void testParsingBoundedValue()
+    {
+        assertBoundedThreadsCount("3", "1", "1", 1);
+        assertBoundedThreadsCount("256C", "4", "16", 16);
+        assertBoundedThreadsCount("3", "4", "16", 4);
+    }
+
+    private void assertThreadsCount(String value, int expected)
+    {
+        ThreadCount threadsCount = ThreadCount.valueOf(value);
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:65]: Nit: `threadCount`

File: src/test/java/io/airlift/units/TestThreadsCount.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+        assertThat(threadsCount).isEqualTo(ThreadCount.exactValueOf(expected));
+        assertThat(threadsCount.getThreadCount()).isEqualTo(expected);
+    }
+
+    private void assertBoundedThreadsCount(String value, String min, String max, int expected)
+    {
+        ThreadCount threadsCount = ThreadCount.boundedValueOf(value, min, max);
```
Comment (electrum) [inline src/test/java/io/airlift/units/TestThreadsCount.java:72]: Same here

**Review by electrum — APPROVED** [review]
Let's drop the `@PowerOfTwo` validator. If we're going to do the clamping in the caller, then it doesn't make sense to have this validator. Also, it wouldn't work with `ThreadCount` since that isn't an integer.

File: pom.xml [hunk 22]
```diff
@@ -71,14 +71,16 @@
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>testing</artifactId>
-            <version>218</version>
+            <!-- 219 is the last version targeted for JDK 11 -->
```
Comment (electrum) [inline pom.xml:74]: @dain @martint should we update units to JDK 17?

Comment (wendigo) [reply pom.xml:74]: It's used in JDBC and client

Comment (wendigo) [reply pom.xml:74]: Already a part of https://github.com/airlift/units/commit/9336b77ca379c12986a2b7051c280c7b25a013b2

**Comment by wendigo** [comment]
@electrum PTAL. Rebased and addressed all comments. I've added `@MinThreadCount` and `@MaxThreadCount` validators
